### DISCUSSION TITLE
How to get a custom DocumentProcessor or PostProcessor to work

### DISCUSSION THREAD

**Discussion by michfuchs** [dotnet/docfx#9239] [Q&A] [answered]
One appraoch to solve #9238 seems to be implementing a custom PostProcessor, as described in the [Documentation](https://dotnet.github.io/docfx/tutorial/howto_add_a_customized_post_processor.html). 

![image](https://github.com/dotnet/docfx/assets/9607008/a6e96aaa-bf7c-4097-b3e9-9f43f9c8a4cc)

However, I don't seem to be able to load my post processor. 

My docfx.json configuration for the post processors looks like this:

`"postProcessors": [ "ExtractSearchIndex", "MyPostProcessor" ]`

Since ExtractSearchIndex works, I'm confident it's in the right spot.

Step 4 in the documentation specifies where the built dll of the assembly containing the preprocessor should be copied to:
![image](https://github.com/dotnet/docfx/assets/9607008/8cfbcd13-93dd-42da-be11-90ea723a2a01)

I tried several folder options, none seem to work:
![image](https://github.com/dotnet/docfx/assets/9607008/5251e4ed-bd22-4690-bb04-d5471bebd1f4)
![image](https://github.com/dotnet/docfx/assets/9607008/93d936d4-8af1-464d-a36b-ada63e410739)
![image](https://github.com/dotnet/docfx/assets/9607008/e1390d68-1678-4aa2-8dd3-a2ac7d93ecdb)
![image](https://github.com/dotnet/docfx/assets/9607008/d71beab1-00e3-42de-8a2e-c4a9dd683f0d)

I am consistently getting the following error:

```
1>Searching custom plugins in directory ****\802_Documentation\templates\my-template\plugins...
1>6 plug-in(s) loaded.
1>Post processor ExtractSearchIndex loaded.
1>EXEC : warning : Can't import: MyPostProcessor, System.Composition.Hosting.CompositionFailedException: No export was found for the contract 'IPostProcessor "MyPostProcessor"'.
1>   at System.Composition.CompositionContext.GetExport(CompositionContract contract)
1>   at System.Composition.CompositionContext.GetExport(Type exportType, String contractName)
1>   at Docfx.Common.CompositionContainer.GetExport(CompositionHost container, Type type, String name) in /_/src/Docfx.Build.Engine/CompositionContainer.cs:line 41
1>EXEC : warning : Can't find the post processor: MyPostProcessor
```
It was always 6 plug-ins, even before I added the processor, so the plug-in is definitely not picked up.

**Comment by filzrev** [comment]
> One approach to solve `#9238` seems to be implementing a custom PostProcessor, as described in the [Documentation](https://dotnet.github.io/docfx/tutorial/howto_add_a_customized_post_processor.html).

It seems [Document](https://dotnet.github.io/docfx/tutorial/howto_add_a_customized_post_processor.html#step4-build-your-project-and-copy-the-output-dll-files-to) is a little outdated and needs updating.

Current docfx implementation load plugin DLLs from following locations. (it's described at [this page](https://dotnet.github.io/docfx/tutorial/howto_build_your_own_type_of_documentation_with_custom_plug-in.html#enable-plug-in))

> Global: the Docfx executable directory.
> Non-global: a folder you create with the name plugins under a template folder. 

And the message `6 plug-in(s) loaded.` is a little misleading.
It simply prints the number of `DocumentProcessor`, not the number of plugins.
https://github.com/dotnet/docfx/blob/f205a44fcafb66345624d9e6d1523f83989c7d10/src/Docfx.Build.Engine/DocumentBuilder.cs#L40-L44

**Reply by michfuchs** [reply]
I also tried to put it right next to the executable, same error. The build log would suggest another folder way down from docfx.exe, but that one also doesn't work. 
![image](https://github.com/dotnet/docfx/assets/9607008/bed34662-c612-42ec-a4de-c46a872a2ed9)

Strange thing is, if I serve the site up, docfx is actually locking the file so that I can't delete it. So it does seem like the file is being checked, but the error message still stands as it is, and no code seems to be executed in the processor.
![image](https://github.com/dotnet/docfx/assets/9607008/daa545f0-4644-4ae8-8cbe-11f0ea98e9f6)

**Reply by filzrev** [reply]
I've detected another obsoleted documentation issue.
https://dotnet.github.io/docfx/tutorial/howto_add_a_customized_post_processor.html#step0-preparation

If your plugin project referencing `Microsoft.Composition` NuGet package.
It should be removed. And use `System.Composition` namespace instead.

As far as confirmed by your screenshot information.
Plugin DLL seems be successfully loaded. But exported interface failed to be imported by `CompositionContainer`.

The situation makes sense. If your plugin using `Microsoft.Composition` version.

**Reply by michfuchs** [reply]
Thanks for your answer. Unfortunately that's not it, I already have System.Composition package referenced.

![image](https://github.com/dotnet/docfx/assets/9607008/7e24cef8-b96a-411f-a48b-e15e9f7ffa76)

**Comment by filzrev** [comment, marked as answer]
What's version .NET global tools installed (e.g. `2.70.3`)

It seems that problems are occurred when custom plugin referencing newer version of `Docfx.Plugins` (`2.70.4`) .
Try to update docfx version or downgrade `Docfx.Plugins` NuGet package.

**Reply by michfuchs** [reply]
Updated the tool to 2.70.4 and it works. Thank you!
Maybe you could check this in the build and issue a warning, if versions need to be equal to the last digit.

**Reply by filzrev** [reply]
Currently plugin DLL loading failure is reported as `Verbose` level message. so it's not shown by default.

> Scanning assembly file C:\Projects\docfx\samples\seed\template\plugins\Sample.Docfx.Plugins.dll...
> Skipping file C:\Projects\docfx\samples\seed\template\plugins\Sample.Docfx.Plugins.dll due to load failure: Unable to load one or more of the requested types.
> Could not load file or assembly 'Docfx.Plugins, Version=2.70.4.0, Culture=neutral, PublicKeyToken=null'.

It might be better to change log level to `Warning`.

**Comment by billybooth** [comment]
We were experiencing the same with a post-processor we've been using for some time now after updating docfx:

```cmd
C:\Projects\xxx\EssSharp>docfx --version
2.73.2+8a60af927dcb1a034de840021cc7268c309f9d17
```

After updating all packages, we ended up here:
```cmd
C:\Projects\xxx\EssSharp.PostProcessor>dotnet list EssSharp.PostProcessor.sln package
Project 'EssSharp.PostProcessor' has the following package references
   [net7.0]:
   Top-level Package                   Requested   Resolved
   > Docfx.Plugins                     2.73.2      2.73.2
   > Newtonsoft.Json                   13.0.3      13.0.3
   > System.Collections.Immutable      8.0.0       8.0.0
   > System.Composition                8.0.0       8.0.0
```

Even when the `Docfx.Plugins` version matched the docfx version (2.73.2), we received an identical error to the above (which was not very helpful) until we downgraded `System.Composition` to version 7.0.0:

```cmd
Searching custom plugins in directory C:\Projects\xxx\EssSharp\docfx\templates\EssSharp\plugins...
warning: Can't import: EssSharpPostProcessor, System.Composition.Hosting.CompositionFailedException: No export was found for the contract 'IPostProcessor "EssSharpPostProcessor"'.
   at System.Composition.CompositionContext.GetExport(CompositionContract contract)
   at System.Composition.CompositionContext.GetExport(Type exportType, String contractName)
   at Docfx.Common.CompositionContainer.GetExport(CompositionHost container, Type type, String name) in /_/src/Docfx.Build/CompositionContainer.cs:line 41
warning: Can't find the post processor: EssSharpPostProcessor
```

Just leaving this here in case it's helpful for anyone else.

Edit: In case it wasn't clear, downgrading `System.Collections.Immutable` and `System.Composition` to version 7.0.0 corrected the issue. It feels (unfortunately) like we are now in a position where we need to pin docfx to version 2.73.2 and, in the future, coordinate all docfx updates with updates to our plugin.
